source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,407 @@
+/*******************************************************************************
+
+    Contains supporting code for enrollment process.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.EnrollmentManager;
+
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.crypto.Schnorr : Pair;
+import agora.common.crypto.Schnorr : Signature;
+import agora.common.crypto.Schnorr : sign;
+import agora.common.Deserializer;
+import agora.common.Hash;
+import agora.common.Serializer;
+import agora.consensus.data.Enrollment;
+import agora.utils.Log;
+
+import d2sqlite3.database;
+import d2sqlite3.library;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+import ocean.util.log.Logger;
+
+import std.algorithm;
+import std.file;
+import std.path;
+import std.stdio;
+
+mixin AddLogger!();
+
+/*******************************************************************************
+
+    Handle enrollment data and manage the validator set(db)
+
+*******************************************************************************/
+
+public class EnrollmentManager
+{
+    /// SQLite db instance
+    private Database db;
+
+    /// Node's key pair
+    private KeyPair node_key_pair;
+
+    /// Random seed
+    private Scalar random_seed_src;
+
+    /// Path to data file on disk
+    private string node_enroll_data_path;
+
+    /// Random key for enrollment
+    private Pair signature_noise;
+
+    /// Enrollment data object
+    private Enrollment* data;
+
+    /// Database path
+    private string db_path;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
+            seed_file_path = path to the data file storing randon seed
+            key_pair = the keypair of the owner node
+
+    ***************************************************************************/
+
+    public this (string db_path, string data_dir, KeyPair key_pair)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
+
+        this.db = Database(db_path);
+
+        // create the table for validator set if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS validator_set " ~
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+
+        // set node's KeyPair object
+        this.node_key_pair = key_pair;
+
+        // load random seed
+        this.node_enroll_data_path = buildPath(data_dir, "node_enroll_data.dat");
+        loadNodeEnrollmentData();
```

Comment:
My comment about mixing abstraction still stands. This is not unittestable. Why not make it part of the database ?

Reply:
The table, or the validator_set is for storing other validators' information, so it does not fits to the information which is just the signature noise of the enrollment data for a node itself. Do you mean another table for storing just one row for the signature noise? I'd like to make a memory storage in the EnrollmentManager in case of an unittest. What do you think about this?

Reply:
I meant another table. There are other options, but this one seems the most straightforward.

Reply:
I think another table is simple and straightforward also. I will make another table named "node_enroll_data".